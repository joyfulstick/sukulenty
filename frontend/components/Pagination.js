import React, { Component } from 'react'
import Head from 'next/head'
import Link from 'next/link'
import PaginationStyles from './styles/PaginationStyles'
import PropTypes from 'prop-types'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import { perPage } from '../config'
import polishPlural from '../lib/polishPurals'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`

class Pagination extends Component {
  render() {
    return (
      <Query query={PAGINATION_QUERY}>
        {({ data, loading }) => {
          if (loading) return <p>Wczytywanie...</p>

          const { count } = data.itemsConnection.aggregate,
            { page } = this.props,
            nbsp = String.fromCharCode(160)

          let pages = Math.ceil(count / perPage)
          pages = pages <= 0 ? 1 : pages
          const counter = `Strona ${page}${nbsp}z${nbsp}${pages}`

          return (
            <PaginationStyles data-test="pagination">
              <Head>
                <title>Sukulenty! - {counter} </title>
              </Head>
              <Link
                prefetch
                href={{
                  pathname: '/items',
                  query: { page: page - 1 },
                }}
              >
                <a aria-disabled={page <= 1} data-test="prev">
                  ⬅️ Poprzednia
                </a>
              </Link>
              <p data-test="total-pages">{counter}</p>
              <p>
                <span>{count} </span>
                {polishPlural('sukulent', 'sukulenty', 'sukulentów')(count)}
              </p>
              <Link
                prefetch
                href={{
                  pathname: '/items',
                  query: { page: page + 1 },
                }}
              >
                <a aria-disabled={page >= pages} data-test="next">
                  Następna ➡️
                </a>
              </Link>
            </PaginationStyles>
          )
        }}
      </Query>
    )
  }
}

Pagination.propTypes = {
  page: PropTypes.number.isRequired,
}

export { Pagination as default, PAGINATION_QUERY }
