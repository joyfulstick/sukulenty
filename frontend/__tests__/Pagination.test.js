import Pagination, { PAGINATION_QUERY } from '../components/Pagination'
import { MockedProvider } from 'react-apollo/test-utils'
import Router from 'next/router'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import wait from 'waait'

Router.router = {
  push() {},
  prefetch() {},
}

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ]
}

describe('<Pagination/>', () => {
  it('displays a loading message', () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>,
    )
    expect(wrapper.text()).toContain('Wczytywanie...')
  })

  it('renders pagination for 44 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(44)}>
        <Pagination page={1} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    expect(wrapper.find('[data-test="total-pages"]').text()).toEqual(
      'Strona 1 z 3',
    )
    const pagination = wrapper.find('div[data-test="pagination"]')
    expect(toJSON(pagination)).toMatchSnapshot()
  })

  it('disables prev button on first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(44)}>
        <Pagination page={1} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    expect(wrapper.find('[data-test="prev"]').prop('aria-disabled')).toEqual(
      true,
    )
    expect(wrapper.find('[data-test="next"]').prop('aria-disabled')).toEqual(
      false,
    )
  })
  it('disables next button on last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(44)}>
        <Pagination page={3} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    expect(wrapper.find('[data-test="prev"]').prop('aria-disabled')).toEqual(
      false,
    )
    expect(wrapper.find('[data-test="next"]').prop('aria-disabled')).toEqual(
      true,
    )
  })
  it('enables all buttons on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(44)}>
        <Pagination page={2} />
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    expect(wrapper.find('[data-test="prev"]').prop('aria-disabled')).toEqual(
      false,
    )
    expect(wrapper.find('[data-test="next"]').prop('aria-disabled')).toEqual(
      false,
    )
  })
})
