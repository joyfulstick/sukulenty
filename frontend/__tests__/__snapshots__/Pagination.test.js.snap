// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 44 items 1`] = `
<div
  className="PaginationStyles-d94h6q-0 cXPzjF"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathname": "/items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      data-test="prev"
      href="/items?page=0"
      onClick={[Function]}
    >
      ⬅️ Poprzednia
    </a>
  </Link>
  <p
    data-test="total-pages"
  >
    Strona 1 z 3
  </p>
  <p>
    <span>
      44
       
    </span>
    sukulenty
  </p>
  <Link
    href={
      Object {
        "pathname": "/items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={false}
      data-test="next"
      href="/items?page=2"
      onClick={[Function]}
    >
      Następna ➡️
    </a>
  </Link>
</div>
`;
